base/logging.h

```diff
@@ -15,7 +15,270 @@
+DEFINE_CHECK_OP_IMPL(GT, > )
+#undef DEFINE_CHECK_OP_IMPL
+
+// In optimized mode, use CheckOpString to hint to compiler that the while
+// condition is unlikely.
+#define CHECK_OP_LOG(name, op, val1, val2)                            \
+  while (::sling::CheckOpString _result =                             \
+             ::sling::name##Impl(                                     \
+                 ::sling::GetReferenceableValue(val1),                \
+                 ::sling::GetReferenceableValue(val2),                \
+                 #val1 " " #op " " #val2))                            \
+    ::sling::LogMessageFatal(__FILE__, __LINE__) << *(_result.str_)
+
+#define CHECK_OP(name, op, val1, val2) CHECK_OP_LOG(name, op, val1, val2)
+
+// CHECK_EQ/NE/...
+#define CHECK_EQ(val1, val2) CHECK_OP(EQ, ==, val1, val2)
+#define CHECK_NE(val1, val2) CHECK_OP(NE, !=, val1, val2)
+#define CHECK_LE(val1, val2) CHECK_OP(LE, <=, val1, val2)
+#define CHECK_LT(val1, val2) CHECK_OP(LT, <, val1, val2)
+#define CHECK_GE(val1, val2) CHECK_OP(GE, >=, val1, val2)
+#define CHECK_GT(val1, val2) CHECK_OP(GT, >, val1, val2)
+#define CHECK_NOTNULL(val)                                 \
+  ::sling::CheckNotNull(__FILE__, __LINE__, \
+                        "'" #val "' Must be non NULL", (val))
+
+#ifndef NDEBUG
+// DCHECK_EQ/NE/...
+#define DCHECK(condition) CHECK(condition)
+#define DCHECK_EQ(val1, val2) CHECK_EQ(val1, val2)
+#define DCHECK_NE(val1, val2) CHECK_NE(val1, val2)
+#define DCHECK_LE(val1, val2) CHECK_LE(val1, val2)
+#define DCHECK_LT(val1, val2) CHECK_LT(val1, val2)
+#define DCHECK_GE(val1, val2) CHECK_GE(val1, val2)
+#define DCHECK_GT(val1, val2) CHECK_GT(val1, val2)
+
+#else
+
+#define DCHECK(condition) while (false && (condition)) LOG(FATAL)
+
+// NDEBUG is defined, so DCHECK_EQ(x, y) and so on do nothing. However, we still
+// want the compiler to parse x and y, because we don't want to lose potentially
+// useful errors and warnings.
+#define _DCHECK_NOP(x, y) \
+  while (false && ((void) (x), (void) (y), 0)) LOG(FATAL)
+
+#define DCHECK_EQ(x, y) _DCHECK_NOP(x, y)
+#define DCHECK_NE(x, y) _DCHECK_NOP(x, y)
+#define DCHECK_LE(x, y) _DCHECK_NOP(x, y)
+#define DCHECK_LT(x, y) _DCHECK_NOP(x, y)
+#define DCHECK_GE(x, y) _DCHECK_NOP(x, y)
+#define DCHECK_GT(x, y) _DCHECK_NOP(x, y)
+
+#endif
+
+// These are for when you don't want a CHECK failure to print a verbose
+// stack trace.  The implementation of CHECK* in this file already doesn't.
+#define QCHECK(condition) CHECK(condition)
+#define QCHECK_EQ(x, y) CHECK_EQ(x, y)
+#define QCHECK_NE(x, y) CHECK_NE(x, y)
+#define QCHECK_LE(x, y) CHECK_LE(x, y)
+#define QCHECK_LT(x, y) CHECK_LT(x, y)
+#define QCHECK_GE(x, y) CHECK_GE(x, y)
+#define QCHECK_GT(x, y) CHECK_GT(x, y)
+
+template <typename T>
+T &&CheckNotNull(const char *file, int line, const char *exprtext, T &&t) {
+  if (t == nullptr) {
+    LogMessageFatal(file, line) << string(exprtext);
+  }
+  return std::forward<T>(t);
+}
+
+#endif
```

Comment:
Optional: Add "// !defined(LOG)" or something like that

Reply:
Done.